### PR TITLE
Implement injectStyled

### PR DESCRIPTION
**Pull request by lttb**
Implement injectStyled from #4 

@kof

### CODE REVIEW

File: src/injectStyled.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import composeClasses from './utils/compose-classes'
+import type {styledType} from './types'
+
+const injectStyled = (styled: styledType) => (InnerComponent: ReactClass<any>) => {
+  styled.mountSheets()
+
+  const {sheets} = styled
+  const {staticSheet, dynamicSheet} = sheets
+
+  const classNames = new Set([
```
Comment (kof) [inline src/injectStyled.js]: Can we avoid Set? It needs a polyfil.

File: src/injectStyled.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    ...Object.keys(dynamicSheet.classes)
+  ])
+
+  const classes = [...classNames]
+    .reduce((acc, name) => ({
+      ...acc,
+      [name]: composeClasses(staticSheet.classes[name], dynamicSheet.classes[name]),
+    }), {})
+
+  return (...props: any) => createElement(InnerComponent, {sheets, classes, ...props})
```
Comment (kof) [inline src/injectStyled.js]: props is always an object, no?

Comment (lttb) [inline src/injectStyled.js]: right, will fix it

Comment (kof) [inline src/injectStyled.js]: I would probably not inject `sheets` at all. What use cases are there anyways? Also in react-jss I inject `sheet` which refs to dynamicSheet if there is one and to static if there is no. Actually I did it because it was already there for the case people still using props.sheet. From my experience there should be no need to use sheet object in react.

Comment (lttb) [inline src/injectStyled.js]: agree, I made it in relation to `injectSheet`, but it really doesn't look necessary

File: src/injectStyled.js (hunk 1 above)
Comment (lttb) [inline src/injectStyled.js]: yeah, we can

File: src/injectStyled.js (hunk 2 above)
Comment (kof) [inline src/injectStyled.js]: If current sheet prop passing in react-jss makes troubles, we might want to remove it from there also.

File: src/injectStyled.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import {createElement} from 'react'
+
+import composeClasses from './utils/compose-classes'
+import type {styledType} from './types'
+
+const injectStyled = (styled: styledType) => (InnerComponent: ReactClass<any>) => {
```
Comment (kof) [inline src/injectStyled.js]: Where does ReactClass come from?

File: .eslintrc [hunk 4]
```diff
@@ -3,3 +3,6 @@ parser: babel-eslint
 
 env:
   jest: true
+
+globals:
+  ReactClass: true
```
Comment (kof) [inline .eslintrc]: Why is this needed, where do we use such a global?

File: src/index.js [hunk 5]
```diff
@@ -1,24 +1,37 @@
...
-type StyledElementPropsType = {
-  classes: Object,
-  children: ?any,
-  className: ?string,
-}
-
-const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {}) => {
-  let sheet
-  let dynamicSheet
+const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {}): styledType => {
```
Comment (kof) [inline src/index.js]: styledType should be capitelized, no?

File: src/injectStyled.js (hunk 3 above)
Comment (lttb) [inline src/injectStyled.js]: This is native flow type

File: src/index.js (hunk 5 above)
Comment (kof) [inline src/index.js]: Also lets use a type styles = Object and put it into types, once we have a fully typed styles object for jss we can easily replace Object with the new one in types.

File: src/types/index.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    // TODO: use types from jss
+    staticSheet: Object,
+    dynamicSheet: Object,
+  },
+  mountSheets: Function,
+  styles: Object
+}
+export type StyledElementAttrsType = {tag: string, styles: Object}
+export type StyledElementType = Function & StyledElementAttrsType
+export type tagOrStyledElementTypeype = string | StyledElementType
```
Comment (kof) [inline src/types/index.js]: should be capitalized

File: src/index.js [hunk 7]
```diff
@@ -1,24 +1,37 @@
...
+  const sheets = {}
   let counter = 0
 
+  const mountSheets = () => {
+    if (!sheets.staticSheet) {
+      sheets.staticSheet = jss.createStyleSheet(baseStyles, {
+        link: true,
```
Comment (kof) [inline src/index.js]: why do you need link: true for static sheet?

File: .eslintrc (hunk 4 above)
Comment (lttb) [inline .eslintrc]: eslint consider flow types and it works with rules like 'no-undef`, so we need to add it as global.

but maybe there is better solution with some plugin etc., I'll take a look

File: src/index.js (hunk 5 above)
Comment (lttb) [inline src/index.js]: 👍

File: README.md [hunk 8]
```diff
@@ -31,16 +31,21 @@ const PrimaryButton = styled(Button, {
...
+    margin: 10,
+    '& $baseButton': {
+      fontSize: 16
+    }
   },
   baseButton: {
-    padding: 10
+    padding: 10,
+    '& + &': {
+      marginLeft; 10
```
Comment (kof) [inline README.md]: typo

File: src/index.js [hunk 9]
```diff
@@ -1,27 +1,46 @@
...
+
 import filterProps from './utils/filter-props'
+import composeClasses from './utils/compose-classes'
+import type {
+  BaseStylesType,
+  ComponentStylesType,
+  StyledType,
+  StyledElementAttrsType,
+  StyledElementType,
+  TagOrStyledElementTypeype,
```
Comment (kof) [inline src/index.js]: typo

File: src/index.js [hunk 10]
```diff
@@ -32,9 +51,8 @@ const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {})
     const staticTag = `${tag}-${++counter}`
 
     return class StyledElement extends PureComponent {
-      static tag = tag
-
-      static styles = elementStyles
+      static tag: string = tag
+      static styles: ComponentStylesType = elementStyles
```
Comment (kof) [inline src/index.js]: just realized: elementStyles should be elementStyle, ownStyles should be ownStyle

Comment (lttb) [inline src/index.js]: right, thanks

File: src/index.js [hunk 11]
```diff
@@ -1,27 +1,46 @@
...
+    }
+  }
+
   const styled = (
-    tagOrStyledElement: tagOrStyledElementTypeype,
-    ownStyles: Object
+    tagOrStyledElement: TagOrStyledElementTypeype,
+    ownStyles: ComponentStylesType
   ): StyledElementType => {
     const {tag, styles}: StyledElementAttrsType = typeof tagOrStyledElement === 'string'
```
Comment (kof) [inline src/index.js]: Lets use `tagName` instead of `tag` everywhere, its more correct and aligned with element.tagName

File: src/index.js [hunk 12]
```diff
@@ -32,9 +51,8 @@ const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {})
     const staticTag = `${tag}-${++counter}`
```
Comment (kof) [inline src/index.js]: I would call it considered previous comment `staticTagName`

File: src/index.js [hunk 13]
```diff
@@ -46,27 +64,17 @@ const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {})
...
-        }
+        mountSheets()
 
-        if (!sheet.getRule(staticTag)) {
-          sheet.addRule(staticTag, elementStyles)
+        if (!sheets.staticSheet.getRule(staticTag)) {
+          sheets.staticSheet.addRule(staticTag, elementStyles)
         }
 
-        if (dynamicStyles && !dynamicSheet.getRule(this.tagScoped)) {
```
Comment (kof) [inline src/index.js]: Instead of tagScoped: `tagId`

Comment (lttb) [inline src/index.js]: maybe `dynamicTagName`?

File: src/index.js [hunk 14]
```diff
@@ -1,108 +1,120 @@
...
+
   const styled = (
-    tagOrStyledElement: tagOrStyledElementTypeype,
-    ownStyles: Object
+    tagOrStyledElement: TagOrStyledElementType,
+    ownStyle: ComponentStyleType
   ): StyledElementType => {
-    const {tag, styles}: StyledElementAttrsType = typeof tagOrStyledElement === 'string'
-      ? {tag: tagOrStyledElement, styles: {}}
+    const {tagName, style}: StyledElementAttrsType = typeof tagOrStyledElement === 'string'
```
Comment (kof) [inline src/index.js]: tagOrStyledElement -> tagNameOrStyledElement

File: src/index.js [hunk 15]
```diff
@@ -1,108 +1,120 @@
...
-  let dynamicSheet
+const createStyled = (
+  jss?: Function = jssDefault
+) => (
+  baseStyles: BaseStylesType = {}
+): StyledType => {
+  const sheets = {}
   let counter = 0
 
+  const getScopedTagName = (tagName: string) => `${tagName}-${++counter}`
```
Comment (kof) [inline src/index.js]: we can actually move it to the utils and call it `generateTagName`, similar to jss https://github.com/cssinjs/jss/blob/master/src/utils/generateClassName.js

File: src/index.js [hunk 16]
```diff
@@ -1,108 +1,120 @@
...
-        if (!sheet.getRule(staticTag)) {
-          sheet.addRule(staticTag, elementStyles)
+        if (!sheets.staticSheet.getRule(staticTagName)) {
+          sheets.staticSheet.addRule(staticTagName, elementStyle)
         }
 
-        if (dynamicStyles && !dynamicSheet.getRule(this.tagScoped)) {
-          dynamicSheet
+        if (dynamicStyle && !sheets.dynamicSheet.getRule(this.dynamicTagName)) {
+          sheets.dynamicSheet
```
Comment (kof) [inline src/index.js]: just realized we don't need `sheets` object any more.

File: src/index.js [hunk 17]
```diff
@@ -1,108 +1,120 @@
...
-      static tag = tag
-
-      static styles = elementStyles
+      static tagName: string = tagName
+      static style: ComponentStyleType = elementStyle
 
       props: StyledElementPropsType
 
-      tagScoped = ''
+      dynamicTagName = ''
```
Comment (kof) [inline src/index.js]: we could set `dynamicTagName = generateTagName(tagName)` here and remove constructor override, right?

File: src/index.js [hunk 18]
```diff
@@ -1,108 +1,120 @@
...
-        if (dynamicStyles) {
-          dynamicSheet.update(this.tagScoped, nextProps)
+        if (dynamicStyle) {
+          sheets.dynamicSheet.update(this.dynamicTagName, nextProps)
         }
       }
 
       componentWillUnmount() {
-        dynamicSheet.deleteRule(this.tagScoped)
+        sheets.dynamicSheet.deleteRule(this.dynamicTagName)
```
Comment (kof) [inline src/index.js]: why do we remove the rule, we could reuse that rule next time the component mounts, no?

File: src/utils/compose-classes.js [hunk 19]
```diff
@@ -0,0 +1 @@
+export default (...args: any) => args.filter(Boolean).join(' ')
```
Comment (kof) [inline src/utils/compose-classes.js]: just realized, why do we use sometimes camel case and sometimes dash separated file names? In JSS I stick to camel case.

File: src/index.js (hunk 18 above)
Comment (lttb) [inline src/index.js]: because we generate new scoped name in constructor, so we can't get and old dynamicTagName for the pervious instance

Comment (lttb) [inline src/index.js]: but in case of https://github.com/cssinjs/styled-jss/pull/6#discussion_r112869372 it would work

File: src/index.js (hunk 17 above)
Comment (lttb) [inline src/index.js]: nope, cause we can call Components with different styles

```jsx
<Button color="pink" />
<Button color="green" />
```

and we need to have a scoped tagName for each instance

File: src/index.js (hunk 16 above)
Comment (lttb) [inline src/index.js]: we need, because we need mutated sheets here too https://github.com/cssinjs/styled-jss/pull/6/files#diff-f336995dd331140b2022485b207c9fc1R10

File: src/index.js (hunk 18 above)
Comment (kof) [inline src/index.js]: why? the instance will be reused in react ....

File: src/index.js (hunk 17 above)
Comment (kof) [inline src/index.js]: right

File: src/index.js (hunk 16 above)
Comment (kof) [inline src/index.js]: do we reassign them somewhere else than in `mountSheets?`? Should evtl mountSheets return them?

File: src/index.js (hunk 18 above)
Comment (lttb) [inline src/index.js]: right, but react will call `constructor` and we will lost it

File: src/index.js (hunk 16 above)
Comment (lttb) [inline src/index.js]: no, you are right, we can just return them.
will fix it
